File: adventofcode-2016/src/main/java/nl/mvdr/adventofcode/adventofcode2016/day18/Part2.java
package nl.mvdr.adventofcode.adventofcode2016.day18;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import nl.mvdr.adventofcode.solver.Solver;

public class Part2 implements Solver {
    
    private static final Logger LOGGER = LoggerFactory.getLogger(Part2.class);

    @Override
    public String solve(String inputfile) {
        
        return new LikeARogueSolver(400_000).solve(inputfile);
    }

    public static void main(String[] args) {
        Part2 instance = new Part2();

        String result = instance.solve("input-day18.txt");

        LOGGER.info(result);
    }
}

File: adventofcode-2025/src/main/kotlin/nl/mvdr/adventofcode/adventofcode2025/day04/Part2.kt
package nl.mvdr.adventofcode.adventofcode2025.day04

import io.github.oshai.kotlinlogging.KotlinLogging
import nl.mvdr.adventofcode.solver.FunctionSolver

private val logger = KotlinLogging.logger{}

fun solvePart2(lines: Sequence<String>): Int {
    var rolls = parseInput(lines)
    val startingRolls = rolls.size

    var done = false
    while (!done) {
        val newRolls = rolls.filter { roll -> 4 <= roll.neighboursIncludingDiagonals().count(rolls::contains) }.toSet()
        done = newRolls == rolls
        rolls = newRolls
    }

    return startingRolls - rolls.size
}

fun main() {
    val result = FunctionSolver(::solvePart2).solve("input-day04.txt")
    logger.info { result }
}
